### PR TITLE
[core] move GetInternalConfig: NodeInfo -> InternalKV

### PR DESCRIPTION
**Pull request by rynewang**
Every time GetInternalConfig reads from table_storage, but it's never mutated. Moves to internal kv as a simple in-mem get (no more read from redis). This itself should slightly update performance. But with #47736 it should improve start up latency a lot in thousand-node clusters. In theory we can remove it all for good, instead just put it as an InternalKV entry. but let's do things step by step.

### CODE REVIEW

**Comment by jjyao** [comment]
> in theory we can remove it all for good, instead just put it as an internalkv entry. but let's do things step by step

Could you elaborate how to remove it all?

File: src/ray/gcs/gcs_server/gcs_kv_manager.h [hunk 1]
```diff
@@ -100,8 +101,10 @@ class InternalKVInterface {
 /// This implementation class of `InternalKVHandler`.
 class GcsInternalKVManager : public rpc::InternalKVHandler {
  public:
-  explicit GcsInternalKVManager(std::unique_ptr<InternalKVInterface> kv_instance)
-      : kv_instance_(std::move(kv_instance)) {}
+  // `raylet_config_list` must outlive this class.
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_kv_manager.h:104]: why don't we use create a copy?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -575,7 +575,8 @@ void GcsServer::InitKVManager() {
     RAY_LOG(FATAL) << "Unexpected storage type! " << storage_type_;
   }
 
-  kv_manager_ = std::make_unique<GcsInternalKVManager>(std::move(instance));
+  kv_manager_ = std::make_unique<GcsInternalKVManager>(std::move(instance),
+                                                       config_.raylet_config_list);
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:579]: We can completely remove `InternalConfigTable`?

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc [hunk 3]
```diff
@@ -141,6 +141,23 @@ void GcsInternalKVManager::HandleInternalKVKeys(
   }
 }
 
+void GcsInternalKVManager::HandleGetInternalConfig(
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_kv_manager.cc:144]: This doesn't need to be an separate rpc, we can just make it a key-value entry with a pre-defined key and just call `Get` to fetch it?

**Comment by rynewang** [comment]
Yeah we can remove InternalConfigTable all for good, instead just use a predefined key to get from internal kv. However as I said that will be a much bigger change (-200 LoC) so I think we can do it in another PR to unblock this important performance gain.

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc [hunk 4]
```diff
@@ -141,6 +141,23 @@ void GcsInternalKVManager::HandleInternalKVKeys(
...
+    rpc::GetInternalConfigRequest request,
+    rpc::GetInternalConfigReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  reply->set_config(raylet_config_list_);
+  GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+  ++counts_[CountType::GET_INTERNAL_CONFIG_REQUEST];
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_kv_manager.cc:150]: Actually I don't think we need to record the count here. instrumented_io_context already records the total calls of `HandleGetInternalConfig`?

File: src/ray/gcs/gcs_server/gcs_kv_manager.h (hunk 1 above)
Comment (jjyao) [reply src/ray/gcs/gcs_server/gcs_kv_manager.h:104]: Can you resolve this?

**Review by jjyao — APPROVED** [review]
We can merge after resolving 2 comments

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc (hunk 4 above)
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_kv_manager.cc:150]: done

File: src/ray/gcs/gcs_server/gcs_kv_manager.h (hunk 1 above)
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_kv_manager.h:104]: changed to copy.

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 2 above)
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_server.cc:579]: Yes let's do it in another PR

File: src/ray/gcs/gcs_server/gcs_kv_manager.h [hunk 5]
```diff
@@ -14,6 +14,7 @@
 
 #pragma once
 #include <memory>
+#include <string_view>
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_kv_manager.h:17]: rm

**Review by jjyao — APPROVED** [review]

**Pull request merged**